### PR TITLE
examples: added cgroup_skb example

### PR DESCRIPTION
**Pull request by mingleiw**
added cgroup_skb example to show how to attach the ebpf program to the socket egress to count the packets.

1. Updated the README.
2. Added cgroup_skb/bpf/cgroup_skb_example.c and examples/cgroup_skb/main.go

### CODE REVIEW

File: examples/README.md [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # eBPF Examples
 
+* Cgroup - Attach a program to the control groups(cgroups).
+  * [cgroup_skb](cgroup_skb/) - Count the egress packets by attaching the program to the socket. It assumes cgroups2 file system was mounted at /sys/fs/cgroup/unified. If this is not true, run `sudo mkdir /mnt/cgroup2 && sudo mount -t cgroup2 none /mnt/cgroup2`.
```
[the remark this answers is not in the document]
Comment (mingleiw) [reply examples/README.md:4]: Thanks, Lorenz. That's good to know. What does your`/proc/mounts` look like? My box is ubuntu 18.04 and /proc/mounts is look like the following:

```
$ cat /proc/mounts | grep cgroup
tmpfs /sys/fs/cgroup tmpfs ro,nosuid,nodev,noexec,mode=755 0 0
cgroup /sys/fs/cgroup/unified cgroup2 rw,nosuid,nodev,noexec,relatime 0 0
cgroup /sys/fs/cgroup/systemd cgroup rw,nosuid,nodev,noexec,relatime,xattr,name=systemd 0 0
cgroup /sys/fs/cgroup/memory cgroup rw,nosuid,nodev,noexec,relatime,memory 0 0
cgroup /sys/fs/cgroup/cpu,cpuacct cgroup rw,nosuid,nodev,noexec,relatime,cpu,cpuacct 0 0
cgroup /sys/fs/cgroup/pids cgroup rw,nosuid,nodev,noexec,relatime,pids 0 0
cgroup /sys/fs/cgroup/hugetlb cgroup rw,nosuid,nodev,noexec,relatime,hugetlb 0 0
cgroup /sys/fs/cgroup/blkio cgroup rw,nosuid,nodev,noexec,relatime,blkio 0 0
cgroup /sys/fs/cgroup/net_cls,net_prio cgroup rw,nosuid,nodev,noexec,relatime,net_cls,net_prio 0 0
cgroup /sys/fs/cgroup/perf_event cgroup rw,nosuid,nodev,noexec,relatime,perf_event 0 0
cgroup /sys/fs/cgroup/freezer cgroup rw,nosuid,nodev,noexec,relatime,freezer 0 0
cgroup /sys/fs/cgroup/cpuset cgroup rw,nosuid,nodev,noexec,relatime,cpuset 0 0
cgroup /sys/fs/cgroup/rdma cgroup rw,nosuid,nodev,noexec,relatime,rdma 0 0
cgroup /sys/fs/cgroup/devices cgroup rw,nosuid,nodev,noexec,relatime,devices 0 0
```

I modified that code a little bit and updated the pull request(check if mount[2] is cgroup2). Can you test it again on your box? Thanks!

File: examples/cgroup_skb/main.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+// $BPF_CLANG and $BPF_CFLAGS are set by the Makefile.
+//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -cc $BPF_CLANG -cflags $BPF_CFLAGS bpf ./bpf/cgroup_skb_example.c -- -I../headers
+
+const (
+	mapKey      uint32 = 0
+	procFS             = "/proc/mounts"
+	bpfProgName        = "count_egress_packets"
+)
+
+var cgroup2Mount = struct {
```
Comment (lmb) [inline examples/cgroup_skb/main.go:34]: For the example its OK to skip the `sync.Once` and memoizing the result of the lookup I think: less code in an example is better.

File: examples/cgroup_skb/main.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+	once sync.Once
+	path string
+	err  error
+}{}
+
+// The /proc/mounts lists the cgroupv2 mount point with the file system type of `cgroup2`
+// that makes it convenient for locating the cgroupv2 file system for different systems.
+func cgroup2Path() (string, error) {
+	cgroup2Mount.once.Do(func() {
+		mounts, err := ioutil.ReadFile(procFS)
```
Comment (lmb) [inline examples/cgroup_skb/main.go:44]: Nit: os.ReadFile, ioutil is deprecated.

File: examples/cgroup_skb/main.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+	"github.com/cilium/ebpf/rlimit"
+)
+
+// $BPF_CLANG and $BPF_CFLAGS are set by the Makefile.
+//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -cc $BPF_CLANG -cflags $BPF_CFLAGS bpf ./bpf/cgroup_skb_example.c -- -I../headers
+
+const (
+	mapKey      uint32 = 0
+	procFS             = "/proc/mounts"
+	bpfProgName        = "count_egress_packets"
```
Comment (lmb) [inline examples/cgroup_skb/main.go:31]: Unused?

File: examples/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (lmb) [reply examples/README.md]: From my recent ubuntu:
```
$ grep cgroup /proc/self/mounts
cgroup2 /sys/fs/cgroup cgroup2 rw,nosuid,nodev,noexec,relatime,nsdelegate,memory_recursiveprot 0 0
```

Works now, thanks!

File: examples/README.md [hunk 5]
```diff
@@ -16,7 +18,7 @@
 
 ```bash
 cd ebpf/examples/
-go run -exec sudo [./kprobe, ./uretprobe, ./ringbuffer, ...]
+go run -exec sudo [./kprobe, ./uretprobe, ./ringbuffer, ./cgroup_skb, ...]
```
Comment (ti-mo) [inline examples/README.md:21]: This extension is not needed, it's just to indicate how the examples can be run.

File: examples/README.md [hunk 6]
```diff
@@ -1,5 +1,7 @@
 # eBPF Examples
 
+* Cgroup - Attach a program to the control groups(cgroups).
+  * [cgroup_skb](cgroup_skb/) - Count the egress packets by attaching the program to the socket. It assumes cgroups2 file system was supported by your systems.
```
Comment (ti-mo) [inline examples/README.md:4]:
```suggestion
  * [cgroup_skb](cgroup_skb/) - Count packets egressing the current cgroup.
```

File: examples/cgroup_skb/main.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
+//go:build linux
+// +build linux
+
+// This program demonstrates attaching an eBPF program to a control group.
+// The eBPF program will be attached as an egress filter for the socket to count the packets.
```
Comment (ti-mo) [inline examples/cgroup_skb/main.go:5]:
```suggestion
// The eBPF program will be attached as an egress filter, receiving an `__sk_buff` pointer for each outgoing packet.
```

File: examples/cgroup_skb/main.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+	}
+	cgroup, err := os.Open(cgroupFS)
+	if err != nil {
+		log.Fatal(err)
+		return
+	}
+	defer cgroup.Close()
+
+	// Link the count_egress_packets program to the cgroup.
+	_, err = link.AttachCgroup(link.CgroupOptions{
```
Comment (ti-mo) [inline examples/cgroup_skb/main.go:100]: The resulting `Link` object should always be defer-closed for good measure.

**Review by ti-mo — CHANGES_REQUESTED** [review]
Thanks! Sorry this took so long to get to. A few small changes and this is good to go. :+1:

**Comment by mingleiw** [comment]
@ti-mo @lmb  I've addressed all the comments. Thank you for the reviews!

**Review by ti-mo — APPROVED** [review]
Rebased and applied a few fixups. Thanks for your contribution! :tada:

**Comment by zhangzhiqiangcs** [comment]
Hi, @ti-mo @mingleiw I just build and run this example, and got a err
```
loading objects: field CountEgressPackets: program count_egress_packets: load program: invalid argument
```
I tryed to print instructs and comes as
```
count_egress_packets:
          ; int count_egress_packets(struct __sk_buff *skb) {
         0: MovImm dst: r1 imm: 0
          ; u32 key      = 0;
         1: StXMemW dst: rfp src: r1 off: -4 imm: 0
         2: MovImm dst: r6 imm: 1
          ; u64 init_val = 1;
         3: StXMemDW dst: rfp src: r6 off: -16 imm: 0
         4: MovReg dst: r2 src: rfp
         5: AddImm dst: r2 imm: -4
          ; u64 *count = bpf_map_lookup_elem(&pkt_count, &key);
         6: LoadMapPtr dst: r1 map: Array(pkt_count)#3
         8: Call FnMapLookupElem
          ; if (!count) {
         9: JNEImm dst: r0 off: 9 imm: 0
        10: MovReg dst: r2 src: rfp
        11: AddImm dst: r2 imm: -4
        12: MovReg dst: r3 src: rfp
        13: AddImm dst: r3 imm: -16
          ; bpf_map_update_elem(&pkt_count, &key, &init_val, BPF_ANY);
        14: LoadMapPtr dst: r1 map: Array(pkt_count)#3
        16: MovImm dst: r4 imm: 0
        17: Call FnMapUpdateElem
        18: JaImm dst: r0 off: 1 imm: 0
          ; __sync_fetch_and_add(count, 1);
        19: StXXAddDW dst: r0 src: r6
          ; }
        20: MovImm dst: r0 imm: 1
        21: Exit
, invalid argument
2023/03/04 15:07:10 loading objects: field CountEgressPackets: program count_egress_packets: load program: invalid argument
```
Do you have any suggestions for the problems. Thanks for any help!

OS Version is Ubuntu 18.04

**Comment by zhangzhiqiangcs** [comment]
sorry to bother. I found the reason is that my kernel version is too low.

**Pull request merged**